### PR TITLE
fuzzing: Add mesh fuzzer

### PR DESCRIPTION
**Pull request by AdamKorcz**
This PR adds a fuzzer for the mesh parsing routines.

Once this fuzzer is merged, I will make the necessary changes on the OSS-fuzz side to run this fuzzer continuously.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[x] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by howardjohn** [comment]
> Once this fuzzer is merged, I will make the necessary changes on the OSS-fuzz side to run this fuzzer continuously.

Is it possible to have all of the logic in istio/istio repo? Then oss-fuzz just calls ./run-fuzz.sh or whatever and never changes?

This seems more sustainable and easier to test

**Comment by AdamKorcz** [comment]
> Is it possible to have all of the logic in istio/istio repo? Then oss-fuzz just calls ./run-fuzz.sh or whatever and never changes?

Absolutely. It is tried and tested by many other integrated projects.
I will set it up in a separate PR once the two fuzzers are integrated so we can include them as well in the same PR.

**Comment by AdamKorcz** [comment]
I see the release note test is failing. Sorry for that, it is my bad, as I did not check that part in the PR message.

I have just updated the initial PR message.

File: tests/fuzz/mesh_fuzzer.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	mn, err := mesh.ParseMeshNetworks(string(data))
+	if err != nil {
+		return 0
+	}
+	mesh.ResolveHostsInNetworksConfig(mn)
+	return 1
+}
+
+func FuzzValidateMeshConfig(data []byte) int {
+	_, _ = mesh.ApplyMeshConfigDefaults(string(data))
```
Comment (howardjohn) [inline tests/fuzz/mesh_fuzzer.go:32]: should we return 0 on error like in FuzzParseMeshNetworks?

File: tests/fuzz/mesh_fuzzer.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import (
+	"istio.io/istio/pkg/config/mesh"
+)
+
+func FuzzParseMeshNetworks(data []byte) int {
+	mn, err := mesh.ParseMeshNetworks(string(data))
+	if err != nil {
+		return 0
+	}
+	mesh.ResolveHostsInNetworksConfig(mn)
```
Comment (howardjohn) [inline tests/fuzz/mesh_fuzzer.go:27]: This is doing a real network DNS lookup. Is that legal/a good idea in a fuzzer?

**Review by howardjohn — COMMENTED** [review]
lgtm overall

File: tests/fuzz/mesh_fuzzer.go (hunk 1 above)
Comment (AdamKorcz) [reply tests/fuzz/mesh_fuzzer.go:27]: I suggest we don't. The return values don't make any noticeable difference in practice.

File: tests/fuzz/mesh_fuzzer.go (hunk 2 above)
Comment (AdamKorcz) [reply tests/fuzz/mesh_fuzzer.go]: Thanks for the heads up. We don't want to do that. I have removed that part.

**Review by howardjohn — APPROVED** [review]

**Pull request merged**